server/src/main/java/org/elasticsearch/common/metrics/Counters.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.metrics;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Simple usage stat counters based on longs. Internally this is a map mapping from String to a CounterMetric.
+ * Calling toNestedMap() will create a nested map, where each dot of the key name will nest deeper. This class allows the
+ * stats producer to not worry about how the map is actually nested.
+ */
+public class Counters implements Writeable {
+
+    private final Map<String, CounterMetric> counters = new ConcurrentHashMap<>();
+
+    public Counters(StreamInput in) throws IOException {
+        int numCounters = in.readVInt();
+        for (int i = 0; i < numCounters; i++) {
+            inc(in.readString(), in.readVLong());
+        }
+    }
+
+    public Counters(String... names) {
```

Comment:
I thought it would be nice to be able to set a counter to 0, to initialize a counter instead of having it missing. I like for example to report that there has been 0 invocations for the health API instead of nothing at all. What do you think?

Reply:
Ah great point! . That's a good use case - especially given the strict definition of `get` that throws exception for missing keys.